docs/hooks.rst

```diff
@@ -1,6 +1,20 @@
 Hooking the Transaction Machinery
 =================================
 
+The :mod:`transaction` machinery allows application developers to register
+two different groups of callbacks to be called, one group before the
+committing the transaction and one group after.
+
+These hooks are **not** designed to be used as replacements for the
+two-phase commit machinery defined by a resource manager (see
+:doc:`resourcemanager`).  In particular, hook functions **must not** raise
+or propagate exceptions.
+
+.. warning::
+
+   Hook functions which *do* raise or propagate exceptions will leave the
+   application in an undefined state.
```

Comment:
The transaction **itself** is actually undamaged by misbehaving hooks:

- Exceptions raised from a before-commit hook don't prevent a subsequent attempt to commit the same transaction.
- Exceptions raised from an after-commit hook are logged and swallowed.

What is broken is the application's expectations about the transaction:  e.g. not all hooks get called, etc.  The point is that it is undefined behavior.